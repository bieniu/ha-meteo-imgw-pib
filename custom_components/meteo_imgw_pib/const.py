"""Constants for the Meteo IMGW-PIB integration."""

from datetime import timedelta

DOMAIN = "meteo_imgw_pib"

ATTRIBUTION = "Data provided by IMGW-PIB"

CONF_STATION_ID = "station_id"

UPDATE_INTERVAL = timedelta(minutes=30)
