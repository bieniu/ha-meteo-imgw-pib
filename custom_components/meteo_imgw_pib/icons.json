{
    "entity": {
        "sensor": {
            "warning": {
                "default": "mdi:alert-outline"
            }
        }
    }
}
