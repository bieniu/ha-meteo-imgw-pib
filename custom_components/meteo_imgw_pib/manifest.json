{
    "domain": "meteo_imgw_pib",
    "name": "Meteo IMGW-PIB",
    "codeowners": [
        "@bieniu"
    ],
    "config_flow": true,
    "documentation": "https://github.com/bieniu/ha-meteo-imgw-pib",
    "integration_type": "service",
    "iot_class": "cloud_polling",
    "issue_tracker": "https://github.com/bieniu/ha-meteo-imgw-pib/issues",
    "requirements": [
        "imgw_pib>=1.3.0"
    ],
    "version": "0.0.0"
}
