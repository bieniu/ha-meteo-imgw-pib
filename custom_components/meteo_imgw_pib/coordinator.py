"""Data Update Coordinator for Meteo IMGW-PIB integration."""

import logging
from dataclasses import dataclass

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.device_registry import DeviceEntryType, DeviceInfo
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
from imgw_pib import ApiError, ImgwPib, WeatherData

from .const import DOMAIN, UPDATE_INTERVAL

_LOGGER = logging.getLogger(__name__)

type MeteoImgwPibConfigEntry = ConfigEntry[MeteoImgwPibData]


class MeteoImgwPibDataUpdateCoordinator(DataUpdateCoordinator[WeatherData]):
    """Class to manage fetching Meteo IMGW-PIB data API."""

    config_entry: MeteoImgwPibConfigEntry

    def __init__(
        self,
        hass: HomeAssistant,
        config_entry: MeteoImgwPibConfigEntry,
        imgwpib: ImgwPib,
        station_id: str,
    ) -> None:
        """Initialize."""
        self.imgwpib = imgwpib
        self.station_id = station_id
        self.device_info = DeviceInfo(
            entry_type=DeviceEntryType.SERVICE,
            identifiers={(DOMAIN, station_id)},
            manufacturer="IMGW-PIB",
            name=f"{imgwpib.weather_stations[station_id]}",
        )

        super().__init__(
            hass,
            _LOGGER,
            config_entry=config_entry,
            name=DOMAIN,
            update_interval=UPDATE_INTERVAL,
        )

    async def _async_update_data(self) -> WeatherData:
        """Update data via internal method."""
        try:
            return await self.imgwpib.get_weather_data()
        except ApiError as err:
            raise UpdateFailed(err) from err


@dataclass
class MeteoImgwPibData:
    """Data for the Meteo IMGW-PIB integration."""

    coordinator: MeteoImgwPibDataUpdateCoordinator
