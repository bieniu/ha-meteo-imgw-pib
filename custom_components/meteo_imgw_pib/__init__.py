"""The Meteo IMGW-PIB integration."""

from __future__ import annotations

from dataclasses import dataclass
import logging

from aiohttp import ClientError
from imgw_pib import ImgwPib
from imgw_pib.exceptions import ApiError

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryNotReady
from homeassistant.helpers.aiohttp_client import async_get_clientsession

from .const import CONF_STATION_ID
from .coordinator import MeteoImgwPibDataUpdateCoordinator

PLATFORMS: list[Platform] = [Platform.SENSOR]

_LOGGER = logging.getLogger(__name__)

type MeteoImgwPibConfigEntry = ConfigEntry[MeteoImgwPibData]


@dataclass
class MeteoImgwPibData:
    """Data for the Meteo IMGW-PIB integration."""

    coordinator: MeteoImgwPibDataUpdateCoordinator


async def async_setup_entry(
    hass: HomeAssistant, entry: MeteoImgwPibConfigEntry
) -> bool:
    """Set up Meteo IMGW-PIB from a config entry."""
    station_id: str = entry.data[CONF_STATION_ID]

    _LOGGER.debug("Using weather station ID: %s", station_id)

    client_session = async_get_clientsession(hass)

    try:
        imgwpib = await ImgwPib.create(
            client_session,
            weather_station_id=station_id,
        )
    except (ClientError, TimeoutError, ApiError) as err:
        raise ConfigEntryNotReady from err

    coordinator = MeteoImgwPibDataUpdateCoordinator(hass, imgwpib, station_id)
    await coordinator.async_config_entry_first_refresh()

    entry.runtime_data = MeteoImgwPibData(coordinator)

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(
    hass: HomeAssistant, entry: MeteoImgwPibConfigEntry
) -> bool:
    """Unload a config entry."""
    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
