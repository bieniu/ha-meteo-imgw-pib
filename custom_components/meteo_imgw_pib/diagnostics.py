"""Diagnostics support for Meteo IMGW-PIB."""

from __future__ import annotations

from dataclasses import asdict
from typing import Any

from homeassistant.core import HomeAssistant

from .coordinator import MeteoImgwPibConfigEntry


async def async_get_config_entry_diagnostics(
    _hass: HomeAssistant, entry: MeteoImgwPibConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    coordinator = entry.runtime_data.coordinator

    return {
        "config_entry_data": entry.as_dict(),
        "weather_data": asdict(coordinator.data),
    }
