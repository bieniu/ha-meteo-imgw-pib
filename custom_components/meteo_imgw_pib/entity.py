"""Define the IMGW-PIB entity."""

from homeassistant.helpers.update_coordinator import CoordinatorEntity

from .const import ATTRIBUTION
from .coordinator import MeteoImgwPibDataUpdateCoordinator


class MeteoImgwPibEntity(CoordinatorEntity[MeteoImgwPibDataUpdateCoordinator]):
    """Define Meteo IMGW-PIB entity."""

    _attr_attribution = ATTRIBUTION
    _attr_has_entity_name = True

    def __init__(
        self,
        coordinator: MeteoImgwPibDataUpdateCoordinator,
    ) -> None:
        """Initialize."""
        super().__init__(coordinator)

        self._attr_device_info = coordinator.device_info
