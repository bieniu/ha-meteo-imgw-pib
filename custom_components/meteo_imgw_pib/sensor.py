"""IMGW-PIB sensor platform."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass

from homeassistant.components.sensor import (
    SensorDeviceClass,
    SensorEntity,
    SensorEntityDescription,
    SensorStateClass,
)
from homeassistant.const import (
    DEGREE,
    PERCENTAGE,
    UnitOfLength,
    UnitOfPressure,
    UnitOfSpeed,
    UnitOfTemperature,
)
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback
from homeassistant.helpers.typing import StateType
from imgw_pib.const import WEATHER_ALERTS_MAP
from imgw_pib.model import WeatherData

from .coordinator import MeteoImgwPibConfigEntry, MeteoImgwPibDataUpdateCoordinator
from .entity import MeteoImgwPibEntity

PARALLEL_UPDATES = 1


@dataclass(frozen=True, kw_only=True)
class MeteoImgwPibSensorEntityDescription(SensorEntityDescription):
    """Meteo IMGW-PIB sensor entity description."""

    value: Callable[[WeatherData], StateType]
    attrs: Callable[[WeatherData], dict[str, StateType]] | None = None


WEATHER_ALERT_DESCRIPTION = MeteoImgwPibSensorEntityDescription(
    key="weather_alert",
    translation_key="weather_alert",
    device_class=SensorDeviceClass.ENUM,
    options=[*WEATHER_ALERTS_MAP.values(), "none"],
    value=lambda data: data.alert.event if data.alert else "none",
    attrs=lambda data: {
        "level": data.alert.level,
        "probability": data.alert.probability,
        "valid_from": data.alert.valid_from,
        "valid_to": data.alert.valid_to,
    }
    if data.alert is not None
    else {},
)
SENSOR_TYPES: tuple[MeteoImgwPibSensorEntityDescription, ...] = (
    MeteoImgwPibSensorEntityDescription(
        key="temperature",
        native_unit_of_measurement=UnitOfTemperature.CELSIUS,
        device_class=SensorDeviceClass.TEMPERATURE,
        state_class=SensorStateClass.MEASUREMENT,
        suggested_display_precision=1,
        value=lambda data: data.temperature.value,
    ),
    MeteoImgwPibSensorEntityDescription(
        key="humidity",
        native_unit_of_measurement=PERCENTAGE,
        device_class=SensorDeviceClass.HUMIDITY,
        state_class=SensorStateClass.MEASUREMENT,
        suggested_display_precision=0,
        value=lambda data: data.humidity.value,
    ),
    MeteoImgwPibSensorEntityDescription(
        key="pressure",
        native_unit_of_measurement=UnitOfPressure.HPA,
        device_class=SensorDeviceClass.PRESSURE,
        state_class=SensorStateClass.MEASUREMENT,
        suggested_display_precision=0,
        value=lambda data: data.pressure.value,
    ),
    MeteoImgwPibSensorEntityDescription(
        key="wind_speed",
        native_unit_of_measurement=UnitOfSpeed.METERS_PER_SECOND,
        device_class=SensorDeviceClass.WIND_SPEED,
        state_class=SensorStateClass.MEASUREMENT,
        suggested_display_precision=0,
        value=lambda data: data.wind_speed.value,
    ),
    MeteoImgwPibSensorEntityDescription(
        key="wind_direction",
        translation_key="wind_direction",
        native_unit_of_measurement=DEGREE,
        device_class=SensorDeviceClass.WIND_DIRECTION,
        state_class=SensorStateClass.MEASUREMENT_ANGLE,
        suggested_display_precision=0,
        value=lambda data: data.wind_direction.value,
        attrs=lambda data: {
            "direction_name": _get_wind_direction(data.wind_direction.value)
        }
        if data.wind_direction.value is not None
        else {},
    ),
    MeteoImgwPibSensorEntityDescription(
        key="precipitation",
        native_unit_of_measurement=UnitOfLength.MILLIMETERS,
        device_class=SensorDeviceClass.PRECIPITATION,
        state_class=SensorStateClass.MEASUREMENT,
        suggested_display_precision=0,
        value=lambda data: data.precipitation.value,
    ),
)


async def async_setup_entry(
    _hass: HomeAssistant,
    entry: MeteoImgwPibConfigEntry,
    async_add_entities: AddConfigEntryEntitiesCallback,
) -> None:
    """Add a Meteo IMGW-PIB sensor entity from a config_entry."""
    coordinator = entry.runtime_data.coordinator

    entities = [
        *(
            MeteoImgwPibSensorEntity(coordinator, description)
            for description in SENSOR_TYPES
            if getattr(coordinator.data, description.key).value is not None
        ),
        MeteoImgwPibSensorEntity(coordinator, WEATHER_ALERT_DESCRIPTION),
    ]

    async_add_entities(entities)


class MeteoImgwPibSensorEntity(MeteoImgwPibEntity, SensorEntity):
    """Define MeteoIMGW-PIB sensor entity."""

    entity_description: MeteoImgwPibSensorEntityDescription

    def __init__(
        self,
        coordinator: MeteoImgwPibDataUpdateCoordinator,
        description: MeteoImgwPibSensorEntityDescription,
    ) -> None:
        """Initialize."""
        super().__init__(coordinator)

        self._attr_unique_id = f"{coordinator.station_id}_{description.key}"
        self.entity_description = description

    @property
    def native_value(self) -> StateType:
        """Return the value reported by the sensor."""
        return self.entity_description.value(self.coordinator.data)

    @property
    def extra_state_attributes(self) -> dict[str, StateType]:
        """Return the state attributes."""
        if self.entity_description.attrs:
            return self.entity_description.attrs(self.coordinator.data)

        return {}


def _get_wind_direction(wind_direction_degree: float) -> str:  # noqa: PLR0911,PLR0912
    """Convert wind direction degree to named direction."""
    if 11.25 <= wind_direction_degree < 33.75:  # noqa: PLR2004
        return "nne"
    if 33.75 <= wind_direction_degree < 56.25:  # noqa: PLR2004
        return "ne"
    if 56.25 <= wind_direction_degree < 78.75:  # noqa: PLR2004
        return "ene"
    if 78.75 <= wind_direction_degree < 101.25:  # noqa: PLR2004
        return "e"
    if 101.25 <= wind_direction_degree < 123.75:  # noqa: PLR2004
        return "ese"
    if 123.75 <= wind_direction_degree < 146.25:  # noqa: PLR2004
        return "se"
    if 146.25 <= wind_direction_degree < 168.75:  # noqa: PLR2004
        return "sse"
    if 168.75 <= wind_direction_degree < 191.25:  # noqa: PLR2004
        return "s"
    if 191.25 <= wind_direction_degree < 213.75:  # noqa: PLR2004
        return "ssw"
    if 213.75 <= wind_direction_degree < 236.25:  # noqa: PLR2004
        return "sw"
    if 236.25 <= wind_direction_degree < 258.75:  # noqa: PLR2004
        return "wsw"
    if 258.75 <= wind_direction_degree < 281.25:  # noqa: PLR2004
        return "w"
    if 281.25 <= wind_direction_degree < 303.75:  # noqa: PLR2004
        return "wnw"
    if 303.75 <= wind_direction_degree < 326.25:  # noqa: PLR2004
        return "nw"
    if 326.25 <= wind_direction_degree < 348.75:  # noqa: PLR2004
        return "nnw"
    return "n"
