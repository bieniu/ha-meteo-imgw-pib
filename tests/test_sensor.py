"""Test sensor of Meteo IMGW-PIB integration."""

from unittest.mock import AsyncMock

from pytest_homeassistant_custom_component.common import (
    HomeAssistant,
    MockConfigEntry,
    er,
)
from syrupy import SnapshotAssertion

from custom_components.meteo_imgw_pib.const import CONF_STATION_ID, DOMAIN


async def test_sensor(
    hass: HomeAssistant, mock_imgw_pib_client: AsyncMock, snapshot: SnapshotAssertion
) -> None:
    """Test sensor."""
    entity_registry = er.async_get(hass)

    entry = MockConfigEntry(domain=DOMAIN, data={CONF_STATION_ID: "12200"})
    entry.add_to_hass(hass)
    await hass.config_entries.async_setup(entry.entry_id)
    await hass.async_block_till_done()

    entity_entries = er.async_entries_for_config_entry(entity_registry, entry.entry_id)

    for entity_entry in entity_entries:
        entity_entry_dict = entity_entry.as_partial_dict
        for item in (
            "area_id",
            "categories",
            "config_entry_id",
            "created_at",
            "device_id",
            "hidden_by",
            "id",
            "labels",
            "modified_at",
        ):
            entity_entry_dict.pop(item)
        assert entity_entry_dict == snapshot(name=f"{entity_entry.entity_id}-entry")

        state = hass.states.get(entity_entry.entity_id)._as_dict
        for item in ("context", "last_changed", "last_reported", "last_updated"):
            state.pop(item)
        assert state == snapshot(name=f"{entity_entry.entity_id}-state")
