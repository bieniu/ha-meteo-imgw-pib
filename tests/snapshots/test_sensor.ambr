# serializer version: 1
# name: test_sensor[sensor.warszawa_humidity-entry]
  dict({
    'disabled_by': None,
    'entity_category': None,
    'entity_id': 'sensor.warszawa_humidity',
    'has_entity_name': True,
    'icon': None,
    'name': None,
    'options': ReadOnlyDict({
      'sensor': ReadOnlyDict({
        'suggested_display_precision': 0,
      }),
    }),
    'original_name': 'Humidity',
    'platform': 'meteo_imgw_pib',
    'translation_key': None,
    'unique_id': '12200_humidity',
  })
# ---
# name: test_sensor[sensor.warszawa_humidity-state]
  dict({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by IMGW-PIB',
      'device_class': 'humidity',
      'friendly_name': 'Warszawa Humidity',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': '%',
    }),
    'entity_id': 'sensor.warszawa_humidity',
    'state': '88',
  })
# ---
# name: test_sensor[sensor.warszawa_precipitation-entry]
  dict({
    'disabled_by': None,
    'entity_category': None,
    'entity_id': 'sensor.warszawa_precipitation',
    'has_entity_name': True,
    'icon': None,
    'name': None,
    'options': ReadOnlyDict({
      'sensor': ReadOnlyDict({
        'suggested_display_precision': 0,
      }),
    }),
    'original_name': 'Precipitation',
    'platform': 'meteo_imgw_pib',
    'translation_key': None,
    'unique_id': '12200_precipitation',
  })
# ---
# name: test_sensor[sensor.warszawa_precipitation-state]
  dict({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by IMGW-PIB',
      'device_class': 'precipitation',
      'friendly_name': 'Warszawa Precipitation',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': <UnitOfLength.MILLIMETERS: 'mm'>,
    }),
    'entity_id': 'sensor.warszawa_precipitation',
    'state': '5',
  })
# ---
# name: test_sensor[sensor.warszawa_pressure-entry]
  dict({
    'disabled_by': None,
    'entity_category': None,
    'entity_id': 'sensor.warszawa_pressure',
    'has_entity_name': True,
    'icon': None,
    'name': None,
    'options': ReadOnlyDict({
      'sensor': ReadOnlyDict({
        'suggested_display_precision': 0,
      }),
    }),
    'original_name': 'Pressure',
    'platform': 'meteo_imgw_pib',
    'translation_key': None,
    'unique_id': '12200_pressure',
  })
# ---
# name: test_sensor[sensor.warszawa_pressure-state]
  dict({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by IMGW-PIB',
      'device_class': 'pressure',
      'friendly_name': 'Warszawa Pressure',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': <UnitOfPressure.HPA: 'hPa'>,
    }),
    'entity_id': 'sensor.warszawa_pressure',
    'state': '1002',
  })
# ---
# name: test_sensor[sensor.warszawa_temperature-entry]
  dict({
    'disabled_by': None,
    'entity_category': None,
    'entity_id': 'sensor.warszawa_temperature',
    'has_entity_name': True,
    'icon': None,
    'name': None,
    'options': ReadOnlyDict({
      'sensor': ReadOnlyDict({
        'suggested_display_precision': 1,
      }),
    }),
    'original_name': 'Temperature',
    'platform': 'meteo_imgw_pib',
    'translation_key': None,
    'unique_id': '12200_temperature',
  })
# ---
# name: test_sensor[sensor.warszawa_temperature-state]
  dict({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by IMGW-PIB',
      'device_class': 'temperature',
      'friendly_name': 'Warszawa Temperature',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': <UnitOfTemperature.CELSIUS: '°C'>,
    }),
    'entity_id': 'sensor.warszawa_temperature',
    'state': '12.5',
  })
# ---
# name: test_sensor[sensor.warszawa_wind_direction-entry]
  dict({
    'disabled_by': None,
    'entity_category': None,
    'entity_id': 'sensor.warszawa_wind_direction',
    'has_entity_name': True,
    'icon': None,
    'name': None,
    'options': ReadOnlyDict({
      'sensor': ReadOnlyDict({
        'suggested_display_precision': 0,
      }),
    }),
    'original_name': 'Wind direction',
    'platform': 'meteo_imgw_pib',
    'translation_key': 'wind_direction',
    'unique_id': '12200_wind_direction',
  })
# ---
# name: test_sensor[sensor.warszawa_wind_direction-state]
  dict({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by IMGW-PIB',
      'friendly_name': 'Warszawa Wind direction',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': '°',
    }),
    'entity_id': 'sensor.warszawa_wind_direction',
    'state': '12',
  })
# ---
# name: test_sensor[sensor.warszawa_wind_speed-entry]
  dict({
    'disabled_by': None,
    'entity_category': None,
    'entity_id': 'sensor.warszawa_wind_speed',
    'has_entity_name': True,
    'icon': None,
    'name': None,
    'options': ReadOnlyDict({
      'sensor': ReadOnlyDict({
        'suggested_display_precision': 0,
      }),
      'sensor.private': ReadOnlyDict({
        'suggested_unit_of_measurement': <UnitOfSpeed.KILOMETERS_PER_HOUR: 'km/h'>,
      }),
    }),
    'original_name': 'Wind speed',
    'platform': 'meteo_imgw_pib',
    'translation_key': None,
    'unique_id': '12200_wind_speed',
  })
# ---
# name: test_sensor[sensor.warszawa_wind_speed-state]
  dict({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by IMGW-PIB',
      'device_class': 'wind_speed',
      'friendly_name': 'Warszawa Wind speed',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': <UnitOfSpeed.KILOMETERS_PER_HOUR: 'km/h'>,
    }),
    'entity_id': 'sensor.warszawa_wind_speed',
    'state': '14.4',
  })
# ---
